Change the state machine used for processing new blocks

abacabadabacaba opened the pull request:
Now a newly added block that is not yet trusted is stored in untrustedHead. After the challenge period is over, it can be moved to head. Fixes #13, #15.

ailisp commented on nearbridge/contracts/NearBridge.sol:
```diff
@@ -44,22 +44,28 @@ contract NearBridge is INearBridge {
-    BlockInfo public backupHead;
+
+    // The most recently added block. May still be in its challenge period, so should not be trusted.
+    BlockInfo public untrustedHead;
```
Great to make a difference between head and untrustedHead!

ailisp reviewed (commented):
Looks good, need bump version in package.json if we want to use this immediately. verify_near_headers test failed because it simply try to submit all blocks during a period, so refused by "not sufficiently new" check, either change test or remove some blocks for now

abacabadabacaba commented:
I fixed verify_near_headers, but it still won't pass cli tests because there is new constructor parameter.

maxzaver reviewed (changes requested):
This logic makes contract more complex by adding an additional public method. Instead we can make `head` and `backupHead` private and add an additional method `getTrustedHead()` which returns either `head` or `backupHead` depending on whether `validAfter` has passed.

abacabadabacaba commented:
> This logic makes contract more complex by adding an additional public method. Instead we can make head and `backupHead` private and add an additional method `getTrustedHead()` which returns either `head` or `backupHead` depending on whether `validAfter` has passed.

I think that the logic is now less complex. There is `head` which is trusted. A new method is needed to move `untrustedHead` to `head` when the challenge period is over without pushing a new block. This also makes this block available to `blockHashes` and `blockMerkleRoots`. It is possible to add an extra method, but it won't make things less complex, because there should be still a way to commit a block so that `NearProver` can use it (it uses `blockMerkleRoots`).

Or I can just make `commitBlock` internal, so a block will only become available when another block is submitted on top of it.

Overall, this code needs a lot of refactoring, but for now I try not to change it too much.

maxzaver commented:
> This also makes this block available to blockHashes and blockMerkleRoots. It is possible to add an extra method, but it won't make things less complex, because there should be still a way to commit a block so that NearProver can use it (it uses blockMerkleRoots).

We could add `trustedMerkleRoots` method that filter the last merkle root if the header has not passed the challenge period yet. By adding `commitBlock` we are delegating complexity to the contract user who now need to call an extra method before being able to use this contract.

> Or I can just make commitBlock internal, so a block will only become available when another block is submitted on top of it.

This will make header that passed challenge period not useful until the next header is submitted. Ideally, we want header to be useful as soon as it passes the challenge period.

abacabadabacaba commented:
> We could add `trustedMerkleRoots` method that filter the last merkle root if the header has not passed the challenge period yet.

That's not a good API design. "Trusted" should be the default (and indeed the only) option.

Anyway, I changed the API to return the last block as soon as the challenge period is over.

maxzaver commented on nearbridge/contracts/NearBridge.sol:
```diff
@@ -105,51 +112,32 @@ contract NearBridge is INearBridge {
 
     function checkBlockProducerSignatureInHead(uint256 signatureIndex) public view returns(bool) {
-        if (head.approvals_after_next[signatureIndex].none) {
+        if (untrustedHead.approvals_after_next[signatureIndex].none) {
```
It looks like we should remove this check since it should never be `none`, because `_checkBp` wouldn't allow such block to be submitted.

abacabadabacaba replied:
Why? We only need signatures from 2/3 of the validators, so some could be missing.

maxzaver replied:
You are right, I did not read the code carefully.

maxzaver commented on nearbridge/contracts/NearBridge.sol:
```diff
@@ -247,70 +246,73 @@ contract NearBridge is INearBridge {
+        destBlock.timestamp = src.inner_lite.timestamp;
+        destBlock.epochId = src.inner_lite.epoch_id;
+        destBlock.nextEpochId = src.inner_lite.next_epoch_id;
+        destBlock.hash = src.hash;
```
FYI this changes the definition of the `hash` field to an actual header hash. I assume it is intended.

abacabadabacaba replied:
Yes, this value was never actually used, so I repurposed this field.

maxzaver commented on nearbridge/contracts/NearBridge.sol:
```diff
@@ -75,10 +81,11 @@ contract NearBridge is INearBridge {
     );
 
-    constructor(Ed25519 ed, uint256 _lockEthAmount, uint256 _lockDuration) public {
+    constructor(Ed25519 ed, uint256 _lockEthAmount, uint256 _lockDuration, uint256 _replaceDuration) public {
```
This change requires changes to https://github.com/near/rainbow-bridge-lib and https://github.com/near/rainbow-bridge-cli (we need to add a new CLI arg). Please create corresponding PRs, without them the bridge will be broken.

abacabadabacaba replied:
I really wish the code to deploy the contracts were in the same repository as the contracts themselves. Currently, I cannot merge this PR because it fails CI.

maxzaver replied:
I agree. CC @Kouprin

maxzaver commented on nearbridge/contracts/NearBridge.sol:
```diff
@@ -44,25 +44,31 @@ contract NearBridge is INearBridge {
+    // Address of the account which submitted the last block.
+    address lastSubmitter;
+    // End of challenge period, or zero if there is no block to be challenged.
+    uint lastValidAt;
```
Please add a comment saying something like: "when it is zero `untrustedHead`, `untrustedHeadIsFromNextEpoch`, `untrustedNextBlockProducers` should be considered to be not set".

maxzaver commented on nearbridge/contracts/NearBridge.sol:
```diff
@@ -247,70 +246,73 @@ contract NearBridge is INearBridge {
+        head = untrustedHead;
+        if (untrustedHeadIsFromNextEpoch) {
+            copyBlockProducers(nextBlockProducers, currentBlockProducers);
+            copyBlockProducers(untrustedNextBlockProducers, nextBlockProducers);
```
Please add a comment saying that if block is from the next epoch it is required to have next block producer so `nextBlockProducers` will never be overridden with empty collection.

maxzaver reviewed (approved):
Looks good. There are some suggestions to add comments. We also need to add PR for rainbow-bridge-cli and rainbow-bridge-lib to make the tests pass. Sorry for the delay.

abacabadabacaba commented:
@nearmax I addressed the comments and also fixed another bug, is everything OK now? Will make a PR for near-bridge-cli and -lib a bit later. I can't merge this PR because of CI.

abacabadabacaba commented:
Created pull requests: near/rainbow-bridge-cli#348, near/rainbow-bridge-lib#14.

maxzaver commented on nearbridge/contracts/NearBridge.sol:
```diff
@@ -200,117 +183,117 @@ contract NearBridge is INearBridge {
         } else {
-            // The new block is from the next epoch.
-            _checkBp(nearBlock, nextBlockProducers);
+            revert("NearBridge: Epoch id of the block is not valid");
```
Was this the additional bug that you fixed? It is hard to see because of the forced push.

abacabadabacaba replied:
The bug that I fixed was that I added a check to `checkBlockProducerSignatureInHead` that `signatureIndex` is valid.

maxzaver replied:
Are you talking about `signatureIndex < untrustedBlockProducers.bpsLength` check? Because it seems `!untrustedApprovals[signatureIndex].none` check does not fix vulnerabilities.

maxzaver reviewed (approved):
Approving, assuming https://github.com/near/rainbow-bridge-sol/pull/20/files?file-filters%5B%5D=.sol#diff-b55dd2aebd3038c032230f5c9710455dR212 is the only non-comment change since previous review.

ailisp reviewed (approved):
Looks good!

maxzaver commented on dist.sh:
```diff
@@ -15,6 +15,7 @@ do
   # Get contract name without extension and without directories.
   contract_name="${filename%.*}"
   node_modules/.bin/truffle-flattener "./contracts/${contract_name}.sol" > "dist/${contract_name}.full.sol"
+  sed -i '/^\/\/ SPDX-License-Identifier:/d' "dist/${contract_name}.full.sol"
```
Could we add a comment explaining what is this for?

abacabadabacaba replied:
This was because of nomiclabs/truffle-flattener#55. Will add a comment.

maxzaver commented on nearbridge/contracts/NearBridge.sol:
```diff
@@ -75,26 +82,27 @@ contract NearBridge is INearBridge {
     );
 
-    constructor(Ed25519 ed, uint256 _lockEthAmount, uint256 _lockDuration) public {
+    constructor(Ed25519 ed, uint256 _lockEthAmount, uint256 _lockDuration, uint256 _replaceDuration) public {
```
This PR might be a good opportunity to make sure we follow Solidity style guide: https://solidity.readthedocs.io/en/v0.6.0/style-guide.html#avoiding-naming-collisions

maxzaver commented on nearbridge/contracts/NearBridge.sol:
```diff
@@ -75,26 +82,27 @@ contract NearBridge is INearBridge {
 
-    function challenge(address payable receiver, uint256 signatureIndex) public {
-        require(block.timestamp < head.validAfter, "Lock period already passed");
+    function challenge(address payable receiver, uint256 signatureIndex) override public {
```
AFAIU `override` is only useful when contract inherits from another contract, not from an interface.

abacabadabacaba replied:
In Solidity 0.6, `override` is _mandatory_ when inheriting from an interface. Without it, it just won't compile.

maxzaver commented on nearbridge/contracts/NearBridge.sol:
```diff
@@ -200,117 +183,137 @@ contract NearBridge is INearBridge {
+            res.currentHeight = head.height;
+            res.nextTimestamp = untrustedHead.timestamp;
+            res.nextValidAt = lastValidAt;
+            res.numBlockProducers =
```
It is unclear why do we not return `numBlockProducers` when there is an untrusted head.

abacabadabacaba replied:
We do return the number of block producers when there is an untrusted head. This number is used by the watchdog to determine the number of signatures to check.

maxzaver commented on nearbridge/contracts/NearBridge.sol:
```diff
@@ -200,117 +183,137 @@ contract NearBridge is INearBridge {
-    function addLightClientBlock(bytes memory data) public payable {
+    struct BridgeState {
+        uint currentHeight;
+        uint nextTimestamp;
```
Could we please add comments explaining each field?

maxzaver commented on nearbridge/contracts/NearBridge.sol:
```diff
@@ -200,117 +183,137 @@ contract NearBridge is INearBridge {
 
-        // 6. If next_bps is not none, sha256(borsh(next_bps)) corresponds to the next_bp_hash in inner_lite.
-        if (!nearBlock.next_bps.none) {
+        // Check that the new block is signed by more than 2/3 of the validators.
```
The numbers in the comments were useful because they corresponded to the spec: https://nomicon.io/ChainSpec/LightClient.html

abacabadabacaba committed:
Bump version, change the state machine used for processing new blocks. Now a newly added block that is not yet trusted is stored in untrustedHead. After the challenge period is over, it can be moved to head. Fixes #13, #15.

maxzaver reviewed (approved):
Looks good

Commit:
CHANGELOG

maxzaver commented:
I have tested this change locally, so I am going to merge it and publish the package.
